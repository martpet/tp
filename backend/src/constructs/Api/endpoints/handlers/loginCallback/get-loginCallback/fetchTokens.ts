import camelcaseKeys from 'camelcase-keys';
import fetch from 'node-fetch';

import { FetchTokensResponse, OauthTokens } from '~/constructs/Api/types';
import { getIdTokenPayload } from '~/constructs/Api/utils';
import { authPaths } from '~/consts';

type FetchTokensProps = {
  code: string;
  codeVerifier: string;
  clientId: string;
  authDomain: string;
  loginCallbackUrl: string;
};

export const fetchTokens = async ({
  code,
  codeVerifier,
  clientId,
  authDomain,
  loginCallbackUrl,
}: FetchTokensProps): Promise<OauthTokens> => {
  const response = await fetch(`https://${authDomain}${authPaths.token}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: new URLSearchParams({
      grant_type: 'authorization_code',
      client_id: clientId,
      code,
      code_verifier: codeVerifier,
      redirect_uri: loginCallbackUrl,
    }),
  });

  const data = (await response.json()) as FetchTokensResponse;

  if ('error' in data) {
    throw new Error(data.error);
  }

  return {
    ...camelcaseKeys(data),
    idTokenPayload: getIdTokenPayload(data.id_token),
  };
};
