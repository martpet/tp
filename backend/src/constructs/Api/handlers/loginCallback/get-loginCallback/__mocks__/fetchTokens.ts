import { OauthTokens } from '../../../../apiTypes';

export const dummyOauthTokens: OauthTokens = {
  accessToken: 'dummyAccessToken',
  refreshToken: 'dummyRefreshToken',
  idToken: 'dummyIdToken',
  expiresIn: 123,
  idTokenPayload: {
    sub: 'dummySub',
    nonce: 'dummyIdTokenNonce',
  },
};

export const fetchTokens = vi.fn().mockResolvedValue(dummyOauthTokens);
