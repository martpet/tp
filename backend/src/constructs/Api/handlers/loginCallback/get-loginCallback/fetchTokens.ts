import camelcaseKeys from 'camelcase-keys';
import fetch from 'node-fetch';

import { authPaths } from '~/consts';

import { FetchTokenResponse, IdTokenPayload, OauthTokens } from '../../../apiTypes';

type FetchTokensProps = {
  code: string;
  codeVerifier: string;
  clientId: string;
  authDomain: string;
  loginCallbackUrl: string;
};

export const fetchTokens = async ({
  code,
  codeVerifier,
  clientId,
  authDomain,
  loginCallbackUrl,
}: FetchTokensProps): Promise<OauthTokens> => {
  const response = await fetch(`https://${authDomain}${authPaths.token}`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
    },
    body: new URLSearchParams({
      grant_type: 'authorization_code',
      client_id: clientId,
      code,
      code_verifier: codeVerifier,
      redirect_uri: loginCallbackUrl,
    }),
  });

  const tokens = (await response.json()) as FetchTokenResponse;

  const idTokenPayload = JSON.parse(
    Buffer.from(tokens.id_token.split('.')[1], 'base64').toString()
  ) as IdTokenPayload;

  return {
    ...camelcaseKeys(tokens),
    idTokenPayload,
  };
};
