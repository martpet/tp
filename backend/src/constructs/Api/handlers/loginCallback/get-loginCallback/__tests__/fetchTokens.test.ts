import camelcaseKeys from 'camelcase-keys';
import fetch, { Response } from 'node-fetch';

import { fetchTokens } from '../fetchTokens';

vi.mock('node-fetch');
vi.mock('camelcase-keys');

const idToken = '.eyJkdW1teV9pZHRva2VuX2tleSI6ImR5bW15X2lkdG9rZW5fdmFsIn0=.';

const args = [
  {
    code: 'dummyCode',
    codeVerifier: 'dummyCodeVerifier',
    clientId: 'dummyClientId',
    authDomain: 'dummyAuthDomain',
    loginCallbackUrl: 'dummyLoginCallbackUrl',
  },
] as const;

beforeEach(() => {
  vi.mocked(fetch).mockResolvedValue({
    json: () => Promise.resolve({ id_token: idToken }),
  } as Response);
});

describe('fetchTokens', () => {
  it('calls "fetch" with correct args', async () => {
    await fetchTokens(...args);
    expect(vi.mocked(fetch).mock.calls).toMatchSnapshot();
  });

  it('calls "camelcaseKeys" with correct args', async () => {
    await fetchTokens(...args);
    expect(vi.mocked(camelcaseKeys).mock.calls).toMatchSnapshot();
  });

  it('resolves with a correct value', () => {
    return expect(fetchTokens(...args)).resolves.toMatchSnapshot();
  });
});
