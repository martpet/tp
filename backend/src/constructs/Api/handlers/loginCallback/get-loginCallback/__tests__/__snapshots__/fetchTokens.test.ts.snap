// Vitest Snapshot v1

exports[`fetchTokens > calls "camelcaseKeys" with correct args 1`] = `
[
  [
    {
      "id_token": ".eyJkdW1teV9pZHRva2VuX2tleSI6ImR5bW15X2lkdG9rZW5fdmFsIn0=.",
    },
  ],
]
`;

exports[`fetchTokens > calls "fetch" with correct args 1`] = `
[
  [
    "https://dummyAuthDomain/oauth2/token",
    {
      "body": URLSearchParams {
        Symbol(query): [
          "grant_type",
          "authorization_code",
          "client_id",
          "dummyClientId",
          "code",
          "dummyCode",
          "code_verifier",
          "dummyCodeVerifier",
          "redirect_uri",
          "dummyLoginCallbackUrl",
        ],
        Symbol(context): null,
      },
      "headers": {
        "Content-Type": "application/x-www-form-urlencoded",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`fetchTokens > resolves with a correct value 1`] = `
{
  "dummyCamelCaseKey": "dummyCamelCaseValue",
  "idTokenPayload": {
    "dummy_idtoken_key": "dymmy_idtoken_val",
  },
}
`;
